pkg/cli/tool_base.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/abcxyz/access-on-demand/apis/v1alpha1"
+	"github.com/abcxyz/access-on-demand/pkg/handler"
+	"github.com/abcxyz/access-on-demand/pkg/requestutil"
+	"github.com/abcxyz/pkg/cli"
+	"github.com/posener/complete/v2/predict"
+)
+
+// ToolBaseCommand is the base command for handling tool requests.
+type ToolBaseCommand struct {
+	cli.BaseCommand
+
+	flagPath string
+
+	flagDebug bool
+
+	// testTool is used for testing only.
+	testTool string
+}
+
+func (c *ToolBaseCommand) Flags() *cli.FlagSet {
+	set := cli.NewFlagSet()
+
+	// Command options
+	f := set.NewSection("COMMAND OPTIONS")
+
+	f.StringVar(&cli.StringVar{
+		Name:    "path",
+		Target:  &c.flagPath,
+		Example: "/path/to/file.yaml",
+		Predict: predict.Files("*"),
+		Usage:   `The path of tool request file, in YAML format.`,
+	})
+
+	f.BoolVar(&cli.BoolVar{
+		Name:    "debug",
+		Target:  &c.flagDebug,
+		Default: false,
+		Usage:   `Turn on debug mode to print command outputs.`,
+	})
+
+	return set
+}
+
+func (c *ToolBaseCommand) setup(ctx context.Context, args []string) (*v1alpha1.ToolRequest, *handler.ToolHandler, error) {
+	f := c.Flags()
+	if err := f.Parse(args); err != nil {
+		return nil, nil, fmt.Errorf("failed to parse flags: %w", err)
+	}
+	args = f.Args()
+	if len(args) > 0 {
+		return nil, nil, fmt.Errorf("unexpected arguments: %q", args)
+	}
+
+	if c.flagPath == "" {
+		return nil, nil, fmt.Errorf("path is required")
+	}
+
+	// Read request from file path.
+	var req v1alpha1.ToolRequest
+	if err := requestutil.ReadRequestFromPath(c.flagPath, &req); err != nil {
+		return nil, nil, fmt.Errorf("failed to read %T: %w", &req, err)
+	}
+
+	if err := v1alpha1.ValidateToolRequest(&req); err != nil {
+		return nil, nil, fmt.Errorf("failed to validate %T: %w", &req, err)
+	}
+
+	opts := []handler.ToolHandlerOption{handler.WithStderr(c.Stderr())}
+	if c.flagDebug {
+		opts = append(opts, handler.WithDebugMode(c.Stdout()))
+	}
+	h := handler.NewToolHandler(ctx, opts...)
```

Comment:
Make an interface of the handler so it can be faked.

Reply:
It will make testing the opts harder though, debug and error outputs.